### PR TITLE
DEI-423: Add dotnetcore build and up/download artifacts steps

### PR DESCRIPTION
**Pull request by nilsguillermin**
## JIRA ticket

[DEI-423](https://tetrascience.atlassian.net/browse/DEI-423)

## Description

Modify `publish.yml` to add dotnetcore, dotnetframework build steps.
## Supporting test data

Successful runs of protocol `common/thermofisher-xcalibur-raw-to-ids:v0.0.3`, consisting of:
1. `common/thermofisher-xcalibur-raw-to-raw-json` (task script which was built by this github action)
2. `common/thermofisher-xcalibur-raw-json-to-ids` (untouched, already existing task script)
org: `tetrascience-nguillerm`
https://tetrascience-uat.com/workflows/9c90e2bb-ef50-4279-8039-ebd9a8b1523d
https://tetrascience-uat.com/workflows/ca134be9-aaef-40e7-b6e5-db860fc5606e
https://tetrascience-uat.com/workflows/df7c45d6-02a3-4439-a87f-9ee642ac6018
https://tetrascience-uat.com/workflows/1dfb0d70-bba4-4ccf-870d-cefcc0eecdae

Successful runs of protocol `common/waters-masslynx-raw-to-ds:v0.0.5`, consisting of:
1. `common/masslynx-windows:v0.0.2` (task script which was built by this github action)
2. `common/masslynx-util:v4.1.1` (untouched, already existing task script)
org: `tetrascience-nguillerm`
https://tetrascience-uat.com/workflows/ea9a8714-4e00-4929-8a04-5b09ebebbfc6
https://tetrascience-uat.com/workflows/a934c5b0-8a90-4b8e-89b2-d82dfdec5efd
https://tetrascience-uat.com/workflows/e51004cf-79bd-4f33-8261-5e85062dd814
https://tetrascience-uat.com/workflows/7683b199-4fec-4432-b425-98a3adde10ed

Confirmed identical output on only matching file pair we had (`71J-E14999-082-A2.raw.zip`)

Successful runs of protocol `common/chromeleon-raw-to-ids:v0.0.5` (to prove non-C# task-scripts are NOT affected), using:
`common/chromeleon-raw-to-ids:v0.0.18`
https://tetrascience-uat.com/workflows/323e5266-8093-4ba0-b2fb-5a2bf53a5584
https://tetrascience-uat.com/workflows/095dec18-5a9b-477b-9549-eb7358336aaf
https://tetrascience-uat.com/workflows/f3d1f654-42cb-4ea6-abc9-d5ff561eb299

## Automated testing

- [ ] Unit tests
- [ ] Integration tests
- [ ] API tests
- [ ] End-To-End tests
- [ ] N/A (please clarify)

### CODE REVIEW

**Review by apjansing — APPROVED** [review]

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -33,15 +33,123 @@ on:
...
+    name: "Check DotNet Build Completed Successfully (if at all)"
+    runs-on: ubuntu-latest
+    needs: [build-dotnet, build-framework]
+    outputs:
+      did-dotnet-build: ${{ steps.did-dotnet-build.outputs.did-dotnet-build }}
+    if: ${{ always() }}
+    steps:
+      - id: did-dotnet-build
+        run: |
+          if [[ ${{ needs.build-dotnet.result }} == 'success' || ${{ needs.build-framework.result }} == 'success' ]]; then
```
Comment (nilsguillermin) [inline .github/workflows/publish.yml:128]: This can probably be simplified to two lines of code.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -33,15 +33,123 @@ on:
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup DotNet 3.1.x
+        uses: actions/setup-dotnet@v3
+        with:
+          dotnet-version: "3.1.x"
+
+      - name: Install dependencies
+        run: dotnet restore dotnetcore/src/ThermoRawFileParser.sln
```
Comment (nilsguillermin) [inline .github/workflows/publish.yml:65]: This hard-coded solution name should be generic.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -33,15 +33,123 @@ on:
...
+
+      - name: Build
+        run: dotnet build --configuration Release --no-restore dotnetcore/src/ThermoRawFileParser.sln
```
Comment (nilsguillermin) [inline .github/workflows/publish.yml:68]: Same here, hard-coded solution name needs to be generic.

Comment (nilsguillermin) [reply .github/workflows/publish.yml]: Fixed.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (nilsguillermin) [reply .github/workflows/publish.yml]: Fixed.

File: .github/workflows/publish.yml (hunk 1 above)
Comment (nilsguillermin) [reply .github/workflows/publish.yml:135]: Bash is too annoying, I'm not going to waste time no this.

**Review by abclay — APPROVED** [review]
Looks good to me.  I have a couple questions:

1. Have you tried using this workflow on a non-dotnet repo to make sure that still works?
2. Will these same changes need to be added to `publish_tag.yml`?

**Comment by nilsguillermin** [comment]
> Looks good to me. I have a couple questions:
> 
> 1. Have you tried using this workflow on a non-dotnet repo to make sure that still works?
> 2. Will these same changes need to be added to `publish_tag.yml`?

1. I have, looks like no issues.
https://tetrascience-uat.com/workflows/323e5266-8093-4ba0-b2fb-5a2bf53a5584
https://tetrascience-uat.com/workflows/095dec18-5a9b-477b-9549-eb7358336aaf
https://tetrascience-uat.com/workflows/f3d1f654-42cb-4ea6-abc9-d5ff561eb299
(ran once with previous version, once with v0.0.5 PR version)
2. Yes, they will.

**Pull request merged**